### PR TITLE
add SpatRaster support in cacheGet and cachePut

### PR DESCRIPTION
**Pull request by tscheypidi**

### CODE REVIEW

File: R/cacheGet.R [hunk 1]
```diff
@@ -35,6 +35,20 @@ cacheGet <- function(prefix, type, args = NULL, graph = NULL, ...) {
     vcat(0, " - corrupt cache file ", basename(fname), "! Continue without cache.")
     return(NULL)
   }
+  if (is.list(x)) {
+    for (elem in c("x", "weight")) {
+      if (is.list(x[[elem]]) && !is.null(x[[elem]]$class) && x[[elem]]$class == "SpatRaster") {
```
Comment (pfuehrlich-pik) [inline R/cacheGet.R]:
```suggestion
      if (is.list(x[[elem]]) && identical(x[[elem]]$class, "SpatRaster")) {
```

File: R/cacheGet.R [hunk 2]
```diff
@@ -35,6 +35,20 @@ cacheGet <- function(prefix, type, args = NULL, graph = NULL, ...) {
...
+        x[[elem]] <- .spatRasterLoad(x[[elem]])
+      }
+    }
+  }
   attr(x, "id") <- fname
   return(x)
 }
+
+.spatRasterLoad <- function(x) {
```
Comment (pfuehrlich-pik) [inline R/cacheGet.R]: I'd suggest inlining this function as it's not used anywhere else and it's short.

File: R/cachePut.R [hunk 3]
```diff
@@ -14,20 +14,43 @@
...
 }
+
+.spatRaster2Cache <- function(x, name, fname) {
+  if (!requireNamespace("terra", quietly = TRUE)) stop("Package `terra` required for caching of SpatRaster objects!")
+  if (length(terra::sources(x)) > 1) {
+    stop("Multiple sources of SpatRaster objects in caching currently not supported!")
+  }
+  if (terra::inMemory(x)) return(x)
+  sourceName <- terra::sources(x)
+  sourceType <- sub("^.*\\.([^.]*)$", "\\1", sourceName)
```
Comment (pfuehrlich-pik) [inline R/cachePut.R]:
```suggestion
  sourceType <- tools::file_ext(sourceName)
```

Comment (pfuehrlich-pik) [reply R/cachePut.R]: might need to add `tools` to `Imports` in `DESCRIPTION`

File: R/cachePut.R [hunk 4]
```diff
@@ -14,20 +14,43 @@
...
+  targetName <- sub("\\..*$", paste0("-", name, ".", sourceType), fname)
+  file.copy(sourceName, targetName)
+  Sys.chmod(targetName, mode = "0666", use_umask = FALSE)
+  return(list(class = "SpatRaster",
+               names = names(x),
+               file = targetName))
```
Comment (pfuehrlich-pik) [inline R/cachePut.R]:
```suggestion
              names = names(x),
              file = targetName))
```

File: README.md [hunk 5]
```diff
@@ -55,7 +55,7 @@ In case of questions / problems please contact Jan Philipp Dietrich <dietrich@pi
 
 To cite package **madrat** in publications use:
 
-Dietrich J, Baumstark L, Wirth S, Giannousakis A, Rodrigues R, Bodirsky B, Kreidenweis U, Klein D, Führlich P (2022). _madrat: May All Data be Reproducible and Transparent (MADRaT)_. doi:10.5281/zenodo.1115490 <https://doi.org/10.5281/zenodo.1115490>, R package version 2.21.0, <https://github.com/pik-piam/madrat>.
+Dietrich J, Baumstark L, Wirth S, Giannousakis A, Rodrigues R, Bodirsky B, Kreidenweis U, Klein D, Führlich P (2022). _madrat: May All Data be Reproducible and Transparent (MADRaT)_. doi: 10.5281/zenodo.1115490 (URL: https://doi.org/10.5281/zenodo.1115490), R package version 2.21.0, <URL: https://github.com/pik-piam/madrat>.
```
Comment (pfuehrlich-pik) [inline README.md]: We still don't know where this is coming from, right? Are you still using R 3.6? :sweat_smile:

Comment (tscheypidi) [reply README.md]: no idea

File: R/cacheGet.R (hunk 2 above)
Comment (tscheypidi) [reply R/cacheGet.R]: done

**Review by pfuehrlich-pik — APPROVED** [review]

**Pull request merged**